### PR TITLE
First working version of JSON Schema validation

### PR DESCRIPTION
**Pull request by benj-zen**
This adds the first working version of JSON Schema validation.

Required dependencies:
```
npm install js-yaml
npm install ajv
```

To run the validation execute
```
node validation/validate.js
```

The schemas are not complete yet but work for the example data.

### CODE REVIEW

File: validation/validate.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        rootDir => fs.readdirSync(rootDir).every(
+            schemeDir => !fs.statSync(path.join(rootDir, schemeDir)).isDirectory()
+                || isValidSchemeDir(rootDir, schemeDir)
+        )
+    );
+}
+
+if (!validate()) {
+    process.exit(1);
+}
```
Comment (franziskuskiefer) [inline validation/validate.js]: :upside_down_face:

File: validation/validate.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    return ['encryption', 'signatures'].every(
+        rootDir => fs.readdirSync(rootDir).every(
+            schemeDir => !fs.statSync(path.join(rootDir, schemeDir)).isDirectory()
+                || isValidSchemeDir(rootDir, schemeDir)
+        )
+    );
+}
+
+if (!validate()) {
+    process.exit(1);
```
Comment (franziskuskiefer) [inline validation/validate.js]: Do we get nice error messages before so we know what went wrong?

File: validation/validate.js (hunk 1 above)
Comment (benj-zen) [reply validation/validate.js]: Something wrong with that? :open_mouth:

Comment (benj-zen) [reply validation/validate.js]: Or just the braces?

File: validation/validate.js (hunk 2 above)
Comment (benj-zen) [reply validation/validate.js]: Yes, when iterating over the files their names are displayed and if something goes wrong either the error message of the json schema validator is printed, or an error message by the script (for instance if a file is missing).

**Comment by benj-zen** [comment]
Will add the package.json and configure the CI.

File: validation/validate.js (hunk 1 above)
Comment (franziskuskiefer) [reply validation/validate.js]: All your files are missing new-lines at the end :smile:

Comment (benj-zen) [reply validation/validate.js]: Whoops :sweat_smile:

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+name: JSON Schema Validation
+
+on: [push]
```
Comment (franziskuskiefer) [inline .github/workflows/main.yml]: Let's do this on PRs as well. Then we know if yaml files are correct before merging.

File: tools/validation/validate.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    );
+}
+
+function isValidFile(file, schema) {
+    console.log(file);
+    var isValid = ajv.compile(schema);
+    if (!isValid(yaml.load(fs.readFileSync(file)))) {
+        console.error(isValid.errors);
+        return false;
+    }
```
Comment (franziskuskiefer) [inline tools/validation/validate.js]: Maybe append a log message that validation was successful? I was wondering if the files that were printed are valid or not.

File: .github/workflows/main.yml (hunk 3 above)
Comment (benj-zen) [reply .github/workflows/main.yml]: When a PR is opened you always have to push the commits, or am I missing something.

Comment (franziskuskiefer) [reply .github/workflows/main.yml]: only if you have access to this repo :wink: outside PRs don't involve pushes

File: tools/validation/validate.js (hunk 4 above)
Comment (benj-zen) [reply tools/validation/validate.js]: fixed

File: .github/workflows/main.yml (hunk 3 above)
Comment (benj-zen) [reply .github/workflows/main.yml]: Oh, I didn't consider that.

Comment (benj-zen) [reply .github/workflows/main.yml]: So, just add `pull_request` or something else?

Comment (franziskuskiefer) [reply .github/workflows/main.yml]: yep (here are all triggers explained: https://help.github.com/en/actions/reference/events-that-trigger-workflows)

**Pull request merged**